Nikita prototype discussion

NikitaGallegos opened the pull request:
All right, I spent hours and lost sleep, so I learned a couple of weird things. 

1. Fancy layers cannot be converted into graphmodules, because the vmap function does not work on the Proxy type. Symbolic trace uses the Proxy type to track computations to create nodes for the computation graph. 
2. nn.sequential is not 100% a module. It does not have a name, I think. Whenever I print off all the names in the stitching layer, I get an empty string along with all the name of the submodules. This is weird, because if you use '' as an input name you get a pytorch error saying modules cannot have '' as a name. But apparently nn.sequential can. 
3. Call module uses nn.sequential to create a new name, a name that causes an error if you try to retrieve it with new_from_merge
4. The ability to merge non graphs is cool, but incomplete. Yes, I can input a module and it will create a node for it. However, I do not know the name of this node, thus I cannot put it in the rewire dictionary to actually wire the module into the new merge graph. 

I am actually going to convert the fancyConv2d to a normal conv2D just so I can turn it into a graph and move on to setting everything else up. I realized I do not need regress-able because I will have to design a method to train the downstream which I can just reuse to train the stitching layer.

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+
+#Knob #1.5
+#The selection of the stitching parents from the pretrained models for the given dataset
+imagenet_models = {"resnet50": get_pretrained_model("resnet18"), "resnet101": get_pretrained_model("resnet34"), "ViT": get_pretrained_model("vit_b_32")}
```
calling `get_pretrained_model` loads all the weights into memory and can slow things down. Suggestion: keep these as string names until you later instantiate a particular pair of models.

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+
+#Knob 2
+#The selection of the stitching layer
+stitching_families = {"conv1x1": Conv2d}
```
Even just `Conv2d` should be prefixed by a resize (`Interpolate2d`) stage, no?

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+
+#Knob 3 
+#The selection of the loss function
+loss_functions = {"class_labels": 0, "soft_labels": 1}
```
Not a big deal, but a more standard pattern would be to specify these various options in command line arguments like

    parser.add_argument("--loss", type=str, default="class_labels", choices=["class_labels", "soft_labels"])

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+
+#Dataset root directory, 
+#sShoulnd't it just be datasets? There is no data folder in the src directory?
+#what is number of workers? 
```
If `num_workers` is set to an int > 0, this is how many background threads pytorch will launch for loading data from disk. This adds some parallelism so while the main thread is doing forward and backward passes on a batch, background threads are already starting to prepare the next batch.

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+modelB = modellist["imagenet"]["resnet101"]
+
+#Get the graph trace of our pretrained models 
+traceA = symbolic_trace(modelA)
```
recommend: `GraphModulePlus.new_from_trace(modelA)`

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+
+
+# single conv operations
+#Why do we need to do this? 
```
It's standard-ish practice when fine-tuning models to remove batchnorm.
Also keeping batchnorm layers in the model led past me to some buggy behavior, since "freezing" models just by setting all the parameters `requires_grad` attributes to `False` does not actually freeze the batchnorm parameters (because they are updated without any gradients). Leads to surprising behavior when you're trying to freeze + fine-tune.

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+
+        #use dummy inputs to extract input and out put shape for desired layers
+
+        modelA_ = modelA.extract_subgraph(inputs = ["x"], output=layerA)
```
`inputs` arg is optional here :)

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+
+
+score = {"name": [], "value": [], "ParentA":  [], "ParentB": []}
+for layerA in layersA:
```
aesthetic preference: loop over layers is something I'd do in bash so that one call of the python script = one model trained.

related comment: expect things to fail. Maybe the server gets restarted after doing 1/2 of the analyses. Good practice to (1) check if a model is already trained and (2) quit early (or `continue` the loop if so)

wrongu commented on prototype_experiment.py:
```diff
@@ -0,0 +1,279 @@
+        display_model_graph(modelB_, "DonationB_Fig.png")
+
+        #Stitch the stitching layer to the bottom of A_ subgraph
+        #modelAx = modelA_.new_from_merge({ "DonationA": modelA_, "": stitching_layer}, {"DonationA_"+layerA: [".0"]}, auto_trace = False)
```
let's discuss in person

NikitaGallegos commented:
## ISSUE RESOLVED
I believe I completed a full single stitched experiment with downstream learning. There is only one particularly important issue. Any form of optimizing modelB causes it's accuracy (and always the accuracy of AxB) to plummet to 0. I do not have a clue why. I tested a bunch of things,
- I tried ensuring the stitching layer and downstream learning used the same images and labels, 
- I changed the output from AxB(image) to B(image), to see if the issue was that modelB was training towards a different network, still modelB plummeted to 0
- I changed the parameters being changed by Adam to modelB's parameters, training still plummeted modelB to 0

Any form of adjusting the weights on B using the Adam optimizer will plummet the models performance to 0. I am concerned that the optimizer maybe restarting the weights of modelB in someway causing the huge drop in performance. I expected that modelB's performance would change, but not so drastically, especially when I applied Adam to modelB like training modelB as a regular network.  

Although, I could be crazy, the largest number of batches I ran was 1000. 

I could not figure out what to do, so for now I am going to wait for your advice. I will be moving on to setting up ML_Logger in the meantime.
